src/lib.rs

```diff
@@ -129,6 +128,13 @@ impl<'a> BuildType<'a> {
             }
             // For info about 'kind' values see:
             // https://github.com/rust-lang/cargo/blob/d47a9545db81fe6d7e6c542bc8154f09d0e6c788/src/cargo/core/manifest.rs#L166-L181
+            // The only "Any" artifacts we can support are bins and examples, so let's make sure
+            // no-one slips us a "custom-build" in form of a build.rs in
+            BuildType::Any => artifact
+                .target
+                .kind
+                .iter()
+                .any(|s| s == "bin" || s == "example"),
```

Comment:
@therealprof Sorry for the delay have been rather busy. I know this has already been merged but just took a quick look and noticed a potential problem.

Since `BuildType::Any` now only allows target kinds of `bin` and `example` if someone makes a project for a library they must now explicitly specify `--lib` such as `cargo size --lib` or it will not be able to find the target, previously you could just do `cargo size` and it would use the library.

Like whats in the comment below lib's can be any custom string (since it can be rlib, dylib, etc.) except for a few cases so I think the better option here would be to just exclude `custom-build` since that's the one causing the problem. To prevent other cases becoming a problem it could be good to also exclude `test` and `bench` but may be worth testing to see when they actually show up.

Additionally, it may actually be a good idea to add a build type explicitly for `custom-build` so the tools work on the build script.

Also I noticed that `llvm-size` can actually take multiple binaries but at the moment we always limit to just 1 target so could consider allowing multiple targets for some tools. Will need to check what tools allow this.


I can take a look at making these changes this weekend, hopefully will get some free time I can spend on it.

Reply:
Thanks for looking at it.

I totally agree that this approach may be too pessimistic and could use some tweaking (and maybe some more CI tests, too 😅).

> Since BuildType::Any now only allows target kinds of bin and example if someone makes a project for a library they must now explicitly specify --lib such as cargo size --lib or it will not be able to find the target, previously you could just do cargo size and it would use the library.

I'm not sure whether I would consider auto-matching a lib a bug or a feature at this point as it's certainly a bit esotheric and `--lib` at least clearly expresses intent, but I wouldn't mind allowing it if it doesn't conflict with more common usecases.

> Like whats in the comment below lib's can be any custom string (since it can be rlib, dylib, etc.) except for a few cases so I think the better option here would be to just exclude custom-build since that's the one causing the problem. To prevent other cases becoming a problem it could be good to also exclude test and bench but may be worth testing to see when they actually show up.

I actually started that way but then I ran into weirdness (don't recall exactly which kind of weirdness, sorry) so I turned around and implemented it like it is now.

> Additionally, it may actually be a good idea to add a build type explicitly for custom-build so the tools work on the build script.

I don't know what that would be good for but if we can find a use case I wouldn't mind having it.

> Also I noticed that llvm-size can actually take multiple binaries but at the moment we always limit to just 1 target so could consider allowing multiple targets for some tools. Will need to check what tools allow this.

Oooh, you mean like `cargo size --examples`? That would actually make a killer feature. 🎉

Reply:
> I'm not sure whether I would consider auto-matching a lib a bug or a feature at this point as it's certainly a bit esotheric and `--lib` at least clearly expresses intent, but I wouldn't mind allowing it if it doesn't conflict with more common usecases.

Yeah, maybe the logic needs to be expanded a bit so if the project has only one target it picks that but once we get into cases of multiple targets selecting what would be the most desired target could work better. Whatever creates the least friction to get started.
<br/>

> I actually started that way but then I ran into weirdness (don't recall exactly which kind of weirdness, sorry) so I turned around and implemented it like it is now.

np, I'll poke at it and see what comes up.
<br/>

> I don't know what that would be good for but if we can find a use case I wouldn't mind having it.
> 
> > Also I noticed that llvm-size can actually take multiple binaries but at the moment we always limit to just 1 target so could consider allowing multiple targets for some tools. Will need to check what tools allow this.
> 
> Oooh, you mean like `cargo size --examples`? That would actually make a killer feature. 🎉

yeah that's totally something we could do. I am a little concerned with clarity like if it used every target by default it would probably get our of hand pretty quickly but in the case of bin and lib maybe its fine? not sure yet.
<br/>

I'll create some issues if I run into things to discuss.